src/pool.rs

```diff
@@ -494,55 +523,65 @@ impl ConnectionPool {
+            None => false,
+        }
+    }
 
-        let guard = self.banlist.read();
+    /// Determines if we can try to unban this server
+    pub async fn can_unban(&self, address: &Address) -> bool {
+        // If somehow primary ends up being banned we should return true here
+        if address.role == Role::Primary {
+            return true;
+        }
+
+        // Check if all instances are banned, in that case unban everything
+        let replicas_available = self.addresses[address.shard]
+            .iter()
+            .filter(|addr| addr.role == Role::Replica)
+            .count();
+
+        debug!("Available targets: {}", replicas_available);
 
-        // Everything is banned = nothing is banned.
-        if guard[address.shard].len() == replicas_available {
-            drop(guard);
-            let mut guard = self.banlist.write();
-            guard[address.shard].clear();
-            drop(guard);
+        let read_guard = self.banlist.read();
+        let all_replicas_banned = read_guard[address.shard].len() == replicas_available;
+        drop(read_guard);
+
+        if all_replicas_banned {
+            let mut write_guard = self.banlist.write();
             warn!("Unbanning all replicas.");
-            return false;
+            write_guard[address.shard].clear();
+            drop(write_guard);
+
+            return true;
         }
 
-        // I expect this to miss 99.9999% of the time.
-        match guard[address.shard].get(address) {
-            Some(timestamp) => {
-                let now = chrono::offset::Utc::now().naive_utc();
-                let config = get_config();
-
-                // Ban expired.
-                if now.timestamp() - timestamp.timestamp() > config.general.ban_time {
-                    drop(guard);
-                    warn!("Unbanning {:?}", address);
-                    let mut guard = self.banlist.write();
-                    guard[address.shard].remove(address);
-                    false
-                } else {
-                    debug!("{:?} is banned", address);
-                    true
-                }
-            }
+        // Check if ban time is expired
+        let read_guard = self.banlist.read();
+        let banned_timestamp = match read_guard[address.shard].get(address) {
+            Some(timestamp) => timestamp.clone(),
+            None => return true,
+        };
+        drop(read_guard);
 
-            None => {
-                debug!("{:?} is ok", address);
-                false
-            }
+        let now = chrono::offset::Utc::now().naive_utc();
+
+        if now.timestamp() - banned_timestamp.timestamp() > self.settings.ban_time {
+            warn!("Unbanning {:?}", address);
+            let mut write_guard = self.banlist.write();
+            write_guard[address.shard].remove(address);
+            drop(write_guard);
+
+            true
+        } else {
+            warn!("{:?} is banned", address);
```

Comment:
This will spam the log massively on loaded systems, think 4,000 times per second and more. It should be debug imo.

Reply:
yeah that's fair point.